## Import relevant modules

In [1]:
# Importing the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score, confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Reading the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [4]:
# Reading the CSV file from the Resources folder into a Pandas DataFrame
file_path = Path("Resources/lending_data.csv")
lending_df = pd.read_csv(file_path)

# Reviewing the imported DataFrame
lending_df.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

In [5]:
lending_df.loc[lending_df['loan_status'] == 1 ]

loan_size  interest_rate  borrower_income  debt_to_income  \
75036    18700.0         11.083            84900        0.646643   
75037    17900.0         10.734            81600        0.632353   
75038    18300.0         10.902            83200        0.639423   
75039    18700.0         11.089            85000        0.647059   
75040    17000.0         10.346            78000        0.615385   
...          ...            ...              ...             ...   
77531    19100.0         11.261            86600        0.653580   
77532    17700.0         10.662            80900        0.629172   
77533    17600.0         10.595            80300        0.626401   
77534    16300.0         10.068            75300        0.601594   
77535    15600.0          9.742            72300        0.585062   

       num_of_accounts  derogatory_marks  total_debt  loan_status  
75036               12                 2       54900            1  
75037               11                 2       51600            1  
75038               11                 2       53200            1  
75039               12                 2       55000            1  
75040               10                 2       48000            1  
...                ...               ...         ...          ...  
77531               12                 2       56600            1  
77532               11                 2       50900            1  
77533               11                 2       50300            1  
77534               10                 2       45300            1  
77535                9                 2       42300            1  

[2500 rows x 8 columns]

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [7]:
# Separating the data into labels and features

# Separating the y variable, the labels
y = lending_df['loan_status']

# Separating the X variable, the features
X = lending_df.drop('loan_status', axis = 1)

In [8]:
# Reviewing the y variable Series
y.head()

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

In [9]:
# Reviewing the X variable DataFrame
X.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [10]:
# Checking the balance of our target values
y.value_counts()

loan_status
0    75036
1     2500
Name: count, dtype: int64

### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [15]:
# Importing the train_test_learn module
from sklearn.model_selection import train_test_split

# Spliting the data using train_test_split
# Assigning a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X, 
                                                    y, 
                                                    random_state=1, 
                                                    stratify=y)
print(y_train)

44997    0
13468    0
52163    0
5330     0
9982     0
        ..
28028    0
18051    0
23757    0
6607     0
56309    0
Name: loan_status, Length: 58152, dtype: int64


In [16]:
y_train.value_counts()

loan_status
0    56277
1     1875
Name: count, dtype: int64

In [17]:
# Importing the train_test_learn module
from sklearn.model_selection import train_test_split

# Spliting the data using train_test_split
# Assigning a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X, 
                                                    y, 
                                                    random_state=1)

In [14]:
# Observing the distribution of data in the training target data
y_train.value_counts()

loan_status
0    56271
1     1881
Name: count, dtype: int64

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [9]:
# Importing the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiating the Logistic Regression model
# Assigning a random_state parameter of 1 to the model
logistic_regression_model = LogisticRegression(solver='lbfgs', random_state=1)

# Fitting the model using training data
lr_model = logistic_regression_model.fit(X_train, y_train)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [12]:
# Making a prediction using the testing data
testing_predictions = lr_model.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [16]:
# Printing the balanced_accuracy score of the model
balanced_accuracy = balanced_accuracy_score(y_test, testing_predictions)

print(f"Balanced Accuracy Score: {balanced_accuracy}")

Balanced Accuracy Score: 0.9721077669385362


In [19]:
# Generating a confusion matrix for the model
test_matrix = confusion_matrix(y_test, testing_predictions)
print("Confusion Matrix")
print(test_matrix)

Confusion Matrix
[[18673    86]
 [   32   593]]


In [20]:
# Printing the classification report for the model
testing_report = classification_report(y_test, testing_predictions)
print(testing_report)

              precision    recall  f1-score   support

           0       1.00      1.00      1.00     18759
           1       0.87      0.95      0.91       625

    accuracy                           0.99     19384
   macro avg       0.94      0.97      0.95     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** According to the classification report of the logistic regression model, the healthy loans (`0`) get predicted with 100% accuracy and the high-risk loans (`1`) get predicted at 91% accuracy on average. The accuracy of the overall model is 99%

---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

In [21]:
# Importing the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

# Instantiating the random oversampler model
# # Assign a random_state parameter of 1 to the model
random_oversampler = RandomOverSampler(random_state=1)

# Fitting the original training data to the random_oversampler model
X_resampled, y_resampled = random_oversampler.fit_resample(X_train, y_train)

In [22]:
# Counting the distinct values of the resampled labels data
distinct_values_count = np.unique(y_resampled, return_counts=True)

# Displaying the result
print("Distinct Values and their Counts:")
for value, count in zip(distinct_values_count[0], distinct_values_count[1]):
    print(f"{value}: {count}")

Distinct Values and their Counts:
0: 56277
1: 56277


### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [23]:
# Instantiating the Logistic Regression model
# Assigning a random_state parameter of 1 to the model
logistic_regression_model = LogisticRegression(solver='lbfgs', random_state=1)

# Fitting the model using the resampled training data
lr_resample_model = logistic_regression_model.fit(X_resampled, y_resampled)

# Making a prediction using the testing data
resample_testing_predictions = logistic_regression_model.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [24]:
# Printing the balanced_accuracy score of the model 
resampled_balanced_accuracy = balanced_accuracy_score(y_test, resample_testing_predictions)

print(f"Balanced Accuracy Score with Resampling: {resampled_balanced_accuracy}")

Balanced Accuracy Score with Resampling: 0.9959744975744975


In [26]:
# Generating a confusion matrix for the model
resampled_test_matrix = confusion_matrix(y_test, resample_testing_predictions)
print("Confusion Matrix")
print(resampled_test_matrix)

Confusion Matrix
[[18668    91]
 [    2   623]]


In [27]:
# Printing the classification report for the model
resampled_testing_report = classification_report(y_test, resample_testing_predictions)
print(resampled_testing_report)

              precision    recall  f1-score   support

           0       1.00      1.00      1.00     18759
           1       0.87      1.00      0.93       625

    accuracy                           1.00     19384
   macro avg       0.94      1.00      0.96     19384
weighted avg       1.00      1.00      1.00     19384



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The logistic regression model that is fit with the oversampled data performs slightly better than the previous model in the case of predicting the high-risk loans (`1`). It increased from an overall 91% to 93%. This test is a better prediction of high-risk loans by 2%. The false negative prediction was reduced significantly in this case. The healthy loans (`0`) got predicted at 100% in both cases. 